File: crates/medousa-types/src/capability.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityBindingSummary {
    pub source: String,
    pub reference: String,
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effect_class: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub invoke_via: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityListEntry {
    pub id: String,
    pub title: String,
    pub binding_count: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub domain: String,
    pub has_grapheme: bool,
    pub has_mcp: bool,
    #[serde(default)]
    pub bindings_summary: Vec<CapabilityBindingSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityListResponse {
    pub capabilities: Vec<CapabilityListEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityBinding {
    pub source: String,
    pub reference: String,
    pub priority: u16,
    pub available: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unavailable_reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub invoke_via: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effect_class: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityImplementations {
    #[serde(default)]
    pub grapheme: Vec<CapabilityBinding>,
    #[serde(default)]
    pub mcp: Vec<CapabilityBinding>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityRecommendation {
    pub source: String,
    pub reference: String,
    pub reason: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CapabilityResolveResponse {
    pub capability: String,
    pub title: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub implementations: CapabilityImplementations,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recommended: Option<CapabilityRecommendation>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gateway_unreachable: Option<bool>,
}

File: python/medousa-sdk/src/medousa/types/capability.py
from __future__ import annotations

from medousa.types.daemon_api import MedousaModel


class CapabilityBindingSummary(MedousaModel):
    source: str
    reference: str
    available: bool
    effect_class: str | None = None
    invoke_via: str | None = None


class CapabilityListEntry(MedousaModel):
    id: str
    title: str
    binding_count: int
    description: str | None = None
    domain: str
    has_grapheme: bool
    has_mcp: bool
    bindings_summary: list[CapabilityBindingSummary] = []


class CapabilityListResponse(MedousaModel):
    capabilities: list[CapabilityListEntry]


class CapabilityResolveResponse(MedousaModel):
    capability: str
    title: str
    description: str | None = None
    implementations: dict = {}
    recommended: dict | None = None
    gateway_unreachable: bool | None = None
